main/src/ca/uwaterloo/flix/api/lsp/provider/HoverProvider.scala

```diff
@@ -28,16 +30,46 @@ import scala.annotation.tailrec
 
 object HoverProvider {
 
-  def processHover(uri: String, pos: Position)(implicit index: Index, root: Root, flix: Flix): JObject = {
-    index.query(uri, pos) match {
-      case None => mkNotFound(uri, pos)
+  def processHover(uri: String, pos: Position)(implicit root: Root, flix: Flix): JObject = {
+    var stack: List[AnyRef] = Nil
+    def push(x: AnyRef): Unit = {
+      stack = x :: stack
+    }
+    object HoverConsumer extends Visitor.Consumer {
+      override def consumeType(tpe: Type): Unit = push(tpe)
+      override def consumeExpr(exp: Expr): Unit = push(exp)
+      override def consumeDefSymUse(symUse: DefSymUse): Unit = push(symUse)
+      override def consumeSigSymUse(symUse: SigSymUse): Unit = push(symUse)
+      override def consumeOpSymUse(symUse: OpSymUse): Unit = push(symUse)
+      override def consumeDef(defn: Def): Unit = push(defn)
+      override def consumeSig(sig: Sig): Unit = push(sig)
+      override def consumeOp(op: Op): Unit = push(op)
+      override def consumeFormalParam(fparam: FormalParam): Unit = push(fparam)
+    }
+
+    Visitor.visitRoot(root, HoverConsumer, Visitor.InsideAcceptor(uri, pos))
 
-      case Some(entity) => hoverEntity(entity, uri, pos)
+    stack match {
+      case Nil => mkNotFound(uri, pos)
+      case head :: _ => hoverAny(head, uri, pos)
     }
   }
 
+  private def hoverAny(x: AnyRef, uri: String, pos: Position)(implicit root: Root, flix: Flix): JObject = x match {
+    case tpe: Type => hoverKind(tpe)
+    case exp: Expr => hoverTypeAndEff(exp.tpe, exp.eff, exp.loc)
+    case DefSymUse(sym, loc) => hoverDef(sym, loc)
+    case SigSymUse(sym, loc) => hoverSig(sym, loc)
+    case OpSymUse(symUse, loc) => hoverOp(symUse, loc)
+    case Def(sym, _, _, loc) => hoverDef(sym, loc)
```

Comment:
I think these should be prefixed by their type. Is it Declaration.Def or something else?

Reply:
Good call :)

Reply:
Actually, all of these are top most things within the TypedAst and Ast. So we'd have to prefix them with these

Reply:
We could do that, but I don't know that that would add much towards analysability

Reply:
Prefix with TypedAst. Dont prefix anything from Ast.

Reply:
Like TypedAst.Def, but Expr alone is fine.

Reply:
Wait. Maybe I am just being silly. I suppose there is only one thing called e.g. `Op`?

Reply:
There might be some overlap between Expr.X and TypedAst.X, but I'm not sure. So you might have been onto something